### PR TITLE
Repository Code of Conduct

### PR DESCRIPTION
**Pull request by AffectedArc07**
## What Does This PR Do
Uploads a code of conduct to the repository. This was drafted by me, improved vastly by @Shadeykins (❤️), and has been approved by @Fox-McCloud and @Necaladun 

## Why It's Good For The Repo
Its about time we lay some ground rules on the repo so we have actual policy around things, as theres no real framework for this.

## Changelog
N/A, repo only

### CODE REVIEW

**Comment by Necaladun** [comment]
I hereby approve of this product and or service.

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Important
+
+The Paradise GitHub is not exempt from community and server rules, especially rules 0, 1, and 4. An inability to abide by the rules on the GitHub will result in disciplinary action up to, or including, a repository ban.
```
Comment (S34NW) [inline CODE_OF_CONDUCT.md:3]:
```suggestion
The Paradise GitHub is not exempt from Paradise Station community and server rules, especially rules 0, 1, and 4. An inability to abide by the rules on GitHub will result in disciplinary action up to, or including, a repository ban.
```

**Comment by AffectedArc07** [comment]
Oh and if anyone wants to read this pre-publish
image
Press that, markdown becomes more readable

File: CODE_OF_CONDUCT.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+"Merge-nagging" or similar behaviour is not acceptable. Comments of this nature will result in warnings or an outright ban from the repository depending on general behaviour.
+
+If you exhibit behaviour that's considered to be a net-negative to the community (offensive commentary, repeat violations, constant nagging, personal attacks, etc.), you may be banned from other Paradise services (discord, forums, server, wiki, etc.)
+
+Maintainers reserve the right to permanently revoke access from the repository if your behaviour is considered to be a net negative.
+
+### PR Approval/Objection Info
+
+Heads of Staff and Maintainers have final say on Pull Requests. While thumbs ratios are generally taken into account, they do not dictate whether or not a PR will be merged.
```
Comment (S34NW) [inline CODE_OF_CONDUCT.md:21]:
```suggestion
Heads of Staff and Maintainers have the final say on Pull Requests. While thumbs ratios are generally taken into account, they do not dictate whether or not a PR will be merged.
```

File: CODE_OF_CONDUCT.md [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+After a twenty four hour minimum waiting period, Pull Requests can be merged once they receive approval from both a Head of Staff and a Maintainer. An exception is made for refactors and fixes, which may be merged at a Maintainer's discretion with no waiting period.
```
Comment (S34NW) [inline CODE_OF_CONDUCT.md:23]:
```suggestion
After a twenty-four-hour minimum waiting period, Pull Requests can be merged once they receive approval from both a Head of Staff and a Maintainer. An exception is made for refactors and fixes, which may be merged at a Maintainer's discretion with no waiting period.
```

File: CODE_OF_CONDUCT.md [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+Heads of Staff and Maintainers are not obligated to publicly state their objections to a Pull Request. Attacking or berating either of these roles over an objection will not be tolerated. Additionally, whining over the closure of a PR, the existence of an objection, or similar behaviour, will not be tolerated.
+
+### PR Expectations
+
+All Pull Requests are expected to be tested prior to submission. If a submitted Pull Request fails to pass CI checks, the likelihood of it being merged will be significantly lower. If you can't take the time to compile/test your Pull Request, do not expect a warm reception.
+
+It is expected that contributors discuss larger design changes on the forums prior to coding a Pull Request. The amount of time spent on any given Pull Request is not relevant. Maintainers are not responsible for contributors wasting their time creating features nobody asked for.
+
+Barring highly specific circumstances (such as single line changes, submissions from advanced users, or changes to repo documentation), we will not accept Pull Requests utilising the web editor.
```
Comment (S34NW) [inline CODE_OF_CONDUCT.md:33]:
```suggestion
Barring highly specific circumstances (such as single-line changes, submissions from advanced users, or changes to repo documentation), we will not accept Pull Requests utilising the web editor.
```

**Review by S34NW — APPROVED** [review]
Just some grammar nitpicks, overall very well written and a clear tone.

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (AffectedArc07) [reply CODE_OF_CONDUCT.md:3]: Ill do the other change, but `the GitHub` makes more sense here, since it concerns this repo, not the entire platform GitHub

File: CODE_OF_CONDUCT.md (hunk 3 above)
Comment (AffectedArc07) [reply CODE_OF_CONDUCT.md:23]: Joining these seems weird in my eyes

File: CODE_OF_CONDUCT.md (hunk 4 above)
Comment (AffectedArc07) [reply CODE_OF_CONDUCT.md:33]: Same here

File: CODE_OF_CONDUCT.md [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+"Merge-nagging" or similar behaviour is not acceptable. Comments of this nature will result in warnings or an outright ban from the repository depending on general behaviour.
+
+If you exhibit behaviour that's considered to be a net-negative to the community (offensive commentary, repeat violations, constant nagging, personal attacks, etc.), you may be banned from other Paradise services (discord, forums, server, wiki, etc.)
+
+Maintainers reserve the right to permanently revoke access from the repository if your behaviour is considered to be a net negative.
+
+### PR Approval/Objection Info
+
+Heads of Staff and Maintainers have the final say on Pull Requests. While thumbs ratios are generally taken into account, they do not dictate whether or not a PR will be merged.
```
Comment (SabreML) [inline CODE_OF_CONDUCT.md:21]: "Thumbs ratios" sounds pretty strange IMO, maybe "reaction ratios" would work better?
```suggestion
Heads of Staff and Maintainers have the final say on Pull Requests. While reaction ratios are generally taken into account, they do not dictate whether or not a PR will be merged.
```

File: CODE_OF_CONDUCT.md [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+After a twenty four hour minimum waiting period, Pull Requests can be merged once they receive approval from both a Head of Staff and a Maintainer. An exception is made for refactors and fixes, which may be merged at a Maintainer's discretion with no waiting period.
+
+Heads of Staff and Maintainers are not obligated to publicly state their objections to a Pull Request. Attacking or berating either of these roles over an objection will not be tolerated. Additionally, whining over the closure of a PR, the existence of an objection, or similar behaviour, will not be tolerated.
```
Comment (SabreML) [inline CODE_OF_CONDUCT.md:25]: > Heads of Staff and Maintainers are not obligated to publicly state their objections to a Pull Request.

Since this is becoming officially official, I just want to state for the record that I disagree strongly with this part. (Under most circumstances anyway.)

**Review by SabreML — APPROVED** [review]

File: CODE_OF_CONDUCT.md (hunk 5 above)
Comment (SabreML) [reply CODE_OF_CONDUCT.md:21]: why did it duplicate this

File: CODE_OF_CONDUCT.md [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+Heads of Staff and Maintainers have final say on Pull Requests. While thumbs ratios are generally taken into account, they do not dictate whether or not a PR will be merged.
+
+After a twenty four hour minimum waiting period, Pull Requests can be merged once they receive approval from both a Head of Staff and a Maintainer. An exception is made for refactors and fixes, which may be merged at a Maintainer's discretion with no waiting period.
+
+Heads of Staff and Maintainers are not obligated to publicly state their objections to a Pull Request. Attacking or berating either of these roles over an objection will not be tolerated. Additionally, whining over the closure of a PR, the existence of an objection, or similar behaviour, will not be tolerated.
+
+### PR Expectations
+
+All Pull Requests are expected to be tested prior to submission. If a submitted Pull Request fails to pass CI checks, the likelihood of it being merged will be significantly lower. If you can't take the time to compile/test your Pull Request, do not expect a warm reception.
```
[the remark this answers is not in the document]
Comment (AffectedArc07) [reply CODE_OF_CONDUCT.md:29]: It means stuff like the basic compile test.

We wont penalise for the weird things dreamchecker picks up, such as
```dm
proc/thing()
```
instead of
```dm
/proc/thing()
```

However if the base code fails to even compile, thats grounds for lower merge chance.

[the remark this answers is not in the document]
Comment (SabreML) [reply CODE_OF_CONDUCT.md:29]: Could maybe be made clearer then, but it's probably fine as it is.

File: CODE_OF_CONDUCT.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+### PR Discussion
+
+Comments on Pull Requests should remain relevant to the PR in question and not derail discussions.
+
+Under no circumstances are users to be attacked for their ideas or contributions. While constructive criticism is encouraged, toxicity or general mean-spirited behaviour will not be tolerated. All participants on a given PR or issue are expected to be civil. Failure to do so will result in disciplinary action.
+
+"Merge-nagging" or similar behaviour is not acceptable. Comments of this nature will result in warnings or an outright ban from the repository depending on general behaviour.
+
+If you exhibit behaviour that's considered to be a net-negative to the community (offensive commentary, repeat violations, constant nagging, personal attacks, etc.), you may be banned from other Paradise services (discord, forums, server, wiki, etc.)
```
[the remark this answers is not in the document]
Comment (AffectedArc07) [reply CODE_OF_CONDUCT.md:15]: Quote from my grammar nerd
```
Discord capital D.
The suggestion on the period is just weird and wrong.
The parentheses are encapsulating a parenthetical statement.
Said statements can't exist, formally, outside of a statement.
tl;dr capitalize the D, but moving the period creates a sentence fragment.
```

[the remark this answers is not in the document]
Comment (SabreML) [reply CODE_OF_CONDUCT.md:15]: The sentence is never resolved as it is though, the period is inside the brackets and after "etc." (In which it acts as part of the word), so it never actually acts to close the main sentence.

File: CODE_OF_CONDUCT.md (hunk 6 above)
Comment (AffectedArc07) [reply CODE_OF_CONDUCT.md:25]: The reason we do this is because other maintainers got so much abuse from people nagging the moment we stated an objection. If people could behave, we wouldn't have to do this.

File: CODE_OF_CONDUCT.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+Maintainers reserve the right to permanently revoke access from the repository if your behaviour is considered to be a net negative.
+
+### PR Approval/Objection Info
+
+Heads of Staff and Maintainers have the final say on Pull Requests. While thumbsup/thumbsdown reaction ratios are generally taken into account, they do not dictate whether or not a PR will be merged.
+
+After a twenty four hour minimum waiting period, Pull Requests can be merged once they receive approval from both a Head of Staff and a Maintainer. An exception is made for refactors and fixes, which may be merged at a Maintainer's discretion with no waiting period.
+
+Heads of Staff and Maintainers are not obligated to publicly state their objections to a Pull Request. Attacking or berating either of these roles over an objection will not be tolerated. Additionally, whining over the closure of a PR, the existence of an objection, or similar behaviour, will not be tolerated.
```
Comment (Kyep) [inline CODE_OF_CONDUCT.md:25]: I too strongly disagree with this.

**Review by Kyep — CHANGES_REQUESTED** [review]
While most of this is fine, I object to the "Heads of Staff and Maintainers are not obligated to publicly state their objections to a Pull Request." bit.
I feel like maints/heads explaining their votes should be the norm. If they don't want to do it themselves, they should get someone else (another head/maint, or even a CM) to do it. Whoever they feel can do it.
But contributors shouldn't be left in the dark.
In my opinion, the lack of clear explanations for PR objections is part of the reason why some github members treat maints badly. They get frustrated at the lack of communication, and lash out. 
Having better communication isn't a silver bullet. We do still have to deal with people who are genuinely toxic no matter what. But we shouldn't use toxicity as an excuse for not even explaining what is wrong with a PR.

File: CODE_OF_CONDUCT.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+### PR Expectations
+
+All Pull Requests are expected to be tested prior to submission. If a submitted Pull Request fails to pass CI checks, the likelihood of it being merged will be significantly lower. If you can't take the time to compile/test your Pull Request, do not expect a warm reception.
+
+It is expected that contributors discuss larger design changes on the forums prior to coding a Pull Request. The amount of time spent on any given Pull Request is not relevant. Maintainers are not responsible for contributors wasting their time creating features nobody asked for.
+
+Barring highly specific circumstances (such as single line changes, submissions from advanced users, or changes to repo documentation), we will not accept Pull Requests utilising the web editor.
+
+Pull Requests regarding heavy-handed nerfs, particularly immediately after said mechanic was used, will be tagged with `I ded pls nerf`. A bad experience with a particular mechanic is not a justification for nerfing it.
```
Comment (Kyep) [inline CODE_OF_CONDUCT.md:35]: I suggest re-wording this to: "Don't put up a PR to nerf a mechanic just after you die to it. If you do, it will be given this tag and likely not pass."
That way, the intent is clear, to discourage people from creating PRs nerfing something that they died to.
As written, this guideline arguably might be interpreted to restrict way more then that. It might be argued, for example, to apply even to PRs that someone has a "bad experience" with as a ghost (ie: as a ghost they watched a mechanic that seemed to be totally OP). Even though that (hopefully) is not the intent.

**Comment by AffectedArc07** [comment]
I try to publicly state objections when I can, and I do. 

However, In the less than a *year* I have been maintainer, the amount of complaint's I have had from people over objections and other things is *staggering*. A large part of the maintainer demotivation is caused from the sheer toxic cesspit that is people responding to objections. 

If you get an objection, you can safely assume its todo with design. We *dont* do code objections. Thats called code review. 

> I feel like maints/heads explaining their votes should be the norm. If they don't want to do it themselves, they should get someone else (another head/maint, or even a CM) to do it. Whoever they feel can do it.

It is the norm, and we do it most of the time. There's a difference between "not obligated" and "not going to". It will happen most of the time, but this is for exceptional circumstances, such as ones where we *know* there is going to be backlash if we state why, either due to toxic communities (the vox discord) or PRs that will easily get anyone shot if they speak against them publicly (genderless). I had to face the abuse from both of those, and it was not pleasant. 

> But contributors shouldn't be left in the dark.
> In my opinion, the lack of clear explanations for PR objections is part of the reason why some github members treat maints badly. They get frustrated at the lack of communication, and lash out.

Lack of communication is no excuse for toxicity. If a PR gets closed, you move on and avoid dying on that hill.

**Comment by Kyep** [comment]
I don't expect heads/maints to ALWAYS state their reasons for objecting to a PR. That would be great, but I don't expect it.

My concern is that as-written it doesn't read as "heads/maints should generally explain why they object to a PR, but in exceptional cases, its okay not to, or to ask another staff member to explain it for them". I would be fine with that. It reads as "maints/heads are not expected to explain their objections to a PR". In general. That's what I mean about it setting a bad precedent. I'm worried that as-written it discourages heads/maints from explaining objection reasons in general by simply saying its never required and saying nothing about it being generally a good idea and something that is usually done.

File: CODE_OF_CONDUCT.md (hunk 9 above)
Comment (Shadeykins) [inline CODE_OF_CONDUCT.md:25]: Add `While generally provided,` as a lead-in to the sentence to resolve this ambiguity.

File: CODE_OF_CONDUCT.md (hunk 10 above)
Comment (Shadeykins) [reply CODE_OF_CONDUCT.md:35]: It's intended to include salty ghosts/deadchat players, not just the person who died to it. Immediately issuing a revert to a mechanic the first time it's used is generally bad form. There are usually better ways to go about changing potentially troublesome mechanics than nerfing them into the floor, especially when a contributor hasn't allowed for any time regarding discussion (IE: immediately after a round).

File: CODE_OF_CONDUCT.md (hunk 9 above)
Comment (Shadeykins) [reply CODE_OF_CONDUCT.md:25]: Add "While generally provided," as a lead-in to 25. This resolves the stated ambiguity.

**Review by AffectedArc07 — COMMENTED** [review]
GitHub left me a phantom review box I cant get rid of. I have a pending review with 0 comments.

**Comment by Shadeykins** [comment]
> 
> 
> GitHub left me a phantom review box I cant get rid of. I have a pending review with 0 comments.

It was me. I am the culprit.

File: CODE_OF_CONDUCT.md (hunk 10 above)
Comment (Shadeykins) [reply CODE_OF_CONDUCT.md:35]: Having edited the original, this is the intent.

Ghosts/deadchat/etc. are expected to allow for time for discussion. Immediately posting nerfs the moment you see something doesn't allow for this. Encouraging people to post nerfs the second they think something is unfair is precisely what this is intending to stop.

**Comment by AffectedArc07** [comment]
Tweaked to say that objections will be normally stated

@Kyep

**Comment by Kyep** [comment]
@AffectedArc07 That's enough for me, objection removed.

File: CODE_OF_CONDUCT.md (hunk 9 above)
Comment (SabreML) [reply CODE_OF_CONDUCT.md]: As a final nitpick, "Not obligated" sounds a lot more unfriendly than something like "Not required" does. This probably still works fine though.

**Comment by Fox-McCloud** [comment]
Can definitely get behind this. Toxicity, constant nagging, and politiking is extremely discouraging and disheartening.

**Review by Crazylemon64 — APPROVED** [review]

**Review by variableundefined — APPROVED** [review]
Looks good to go

**Review by Fox-McCloud — APPROVED** [review]
